DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp

```diff
@@ -0,0 +1,90 @@
+#ifndef __AUTORUN_HPP__
+#define __AUTORUN_HPP__
+
+#include <CL/sycl.hpp>
+#include <type_traits>
+
+/*
+This header defines the Autorun kernel utility. This utility is used to
+launch kernels that are submitted before main begins. It is typically used
+to launch kernels that run forever.
+
+Two classes are defined in this header file: Autorun and AutorunForever.
+Autorun creates an autorun kernel that is NOT implicitly wrapped in an infinite
+loop.
+
+AutorunForever creates a kernel and wraps it in a while(1) loop.
+Autorun creates the kernel and does not wrap it in a while(1) loop.
+Usually when using Autorun the user will have the while(1) loop explicitly in
+their code.
+
+The following describes the common template and constructor arguments for both
+the Autorun and AutorunForever.
+
+Template Args:
+  KernelID (optional): the name of the autorun kernel.
+  DeviceSelector: The type of the device selector.
+  KernelFunctor: the kernel functor type.
+Constructor Arguments:
+    device_selector: the SYCL device selector
+    kernel: the user-defined kernel functor.
+            This defines the logic of the autorun kernel.
+*/
+namespace fpga_tools {
+
+namespace detail {
+// Autorun implementation
+template <bool run_forever, typename KernelID>
+struct Autorun_impl {
+  // Constructor with a kernel name
+  template <typename DeviceSelector, typename KernelFunctor>
+  Autorun_impl(DeviceSelector device_selector, KernelFunctor kernel) {
+    // static asserts to ensure KernelFunctor is callable
+    static_assert(std::is_invocable_r_v<void, KernelFunctor>,
+                  "KernelFunctor must be callable with no arguments");
+
+    // create the device queue
+    sycl::queue q{device_selector};
+
+    // submit the user's kernel
+    if constexpr (run_forever) {
+      if constexpr (std::is_same_v<KernelID, void>) {
+        // AutorunForever, kernel name not given
+        q.single_task([=] {
+          while (1) {
+            kernel();
+          }
+        });
+      } else {
+        // AutorunForever, kernel name given
+        q.single_task<KernelID>([=] {
+          while (1) {
+            kernel();
+          }
+        });
+      }
+    } else {
+      // run the kernel as-is, if the user wanted it to run forever they
+      // will write their own explicit while-loop
+      if constexpr (std::is_same_v<KernelID, void>) {
+        // Autorun, kernel name not given
+        q.single_task(kernel);
+      } else {
+        // Autorun, kernel name given
+        q.single_task<KernelID>(kernel);
+      }
+    }
+  }
+};
+}  // namespace detail
+
+// Autorun
+template <typename KernelID = void>
+using Autorun = detail::Autorun_impl<false, KernelID>;
+
+// AutorunForever
+template <typename KernelID = void>
+using AutorunForever = detail::Autorun_impl<true, KernelID>;
+}  // namespace fpga_tools
+
+#endif /* __AUTORUN_HPP__ */
```

Comment:
Missing EOL.

Reply:
Here is the thing. I was always told to have a newline at the end of files. I forget why, but I was always told to. However, clang-format doesn't do it for me and I don't get a warning. I remember in a previous tutorial I had newlines at the end of files and either someone or clang-format made me get rid of them.

Anyways. Long rant for me to say I am not sure what to do. @mdbtucker  ??

Reply:
If clang-format doesn't add the eol, then I don't think you NEED to add it.  The EOL doesn't seem to be part of the Google coding style guide, and I'm not aware of any guidelines in the Code Samples that says we have to do this.

That said, if clang-format doesn't REMOVE the eol, then I think it's at your discretion.  The argument I've found online for why to do this is if you add new lines to the end of the file in a future edit and you don't have an EOL at the end of the file, then the change will show the last line of the file as being edited (to add the EOL).  I think this is pretty weak justification, but it is a justification I guess.

Long story short, I really don't see a good reason to care one way or the other.  I wouldn't fail to approve a PR because of this, but I would also approve a future PR if somebody who does care wants to go in and add that EOL.

Reply:
Nobody seems to care too much about this one, marking it as resolved.  @keryell feel free to reopen if you disagree.

Reply:
At the end of the day, I do not care that much about improving the code of my competitors. :-)
So you can even remove the EOL at the end of all the files to make your code quality worse, even it is sad to have bad code quality on SYCL examples for FPGA in general... :-( @bader

Reply:
We are following the Google coding style for clang-format, which doesn't add newlines to the end of files even though clang-format presents that option.

Reply:
Moreover, based on the C++11 spec (I think this started in C++03), this is no longer necessary with a complian compiler:

C++03 Standard [2.2.1.3] ([page 31](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3690.pdf))

> A source file that is not empty and that does not end in a new-line character, or that ends in a new-line character
immediately preceded by a backslash character before any such splicing takes place, shall be processed
as if an additional new-line character were appended to the file.

I think the only remaining argument is how files that don't do this "play" with other tools (e.g. VIM, cat, ...)? I concede that these other tools may not like this, but we don't think this makes it worth deviating from the Google coding standards for; those people seem to know what they're doing.